torch/csrc/jit/script/exit_transforms.cpp

```diff
@@ -0,0 +1,394 @@
+          handleLoop(node);
+        } break;
+        default:
+          break;
+      }
+      ExitStatus status = getStatus(exit_pair);
+      if (status == WILL) {
+        deleteAfterExitNodes(block, it);
+        break;
+      } else if (status == MIGHT) {
+        if (it != block->nodes().end()) {
+          exit_pair = guardBlockNodes(block, exit_pair, it);
+        }
+        break;
+      }
+    }
+    return exit_pair;
+  }
+
+  void run() {
+    transformExits(graph->block());
+  }
+
+  Value* getUnitValue(const TypePtr& type) {
+    auto maybe_val = unit_values.find(type);
+    if (maybe_val != unit_values.end()) {
+      return maybe_val->second;
+    }
+    auto unit = graph->createUninitialized(type)
+                    ->insertAfter(graph->param_node())
+                    ->output();
+    unit_values[type] = unit;
+    return unit;
+  }
+
+  std::unordered_map<TypePtr, Value*> unit_values;
+  Value* true_val;
+  Value* false_val;
+
+  std::shared_ptr<Graph> graph;
+};
+
+// The Logic for the loop transform simplifies if the BlockExits
+// are converted to LoopExits before running.
+void convertLoopBlockExits(Block* block) {
+  for (auto it = block->nodes().begin(); it != block->nodes().end();) {
+    Node* n = *it;
+    it++;
+    for (Block* b : n->blocks()) {
+      convertLoopBlockExits(b);
+    }
+    if (n->kind() == prim::BlockExit && owningNodeKind(block) == prim::Loop) {
+      auto loop_exit =
+          n->owningGraph()->create(prim::LoopExit, 0)->insertAfter(n);
+      for (auto inp : n->inputs()) {
+        loop_exit->addInput(inp);
+      }
+      n->destroy();
+    }
+  }
+}
+
+// If Node Blocks will only ever have just the one BlockExit,
+// so we can trivially remove them.
+void removeIfNodeExits(Block* block) {
+  for (auto it = block->nodes().begin(), end = block->nodes().end();
+       it != end;) {
+    auto n = *it;
+    it++;
+    for (Block* b : n->blocks()) {
+      removeIfNodeExits(b);
+    }
+    if (n->kind() == prim::BlockExit && owningNodeKind(block) == prim::If) {
+      registerBlockOutputs(block, n->inputs());
+      n->destroy();
+    }
+  }
+}
+
+// This pass takes in a graph where BlockExit & LoopExit exist in the graph
+// and erases them in the graph, correctly setting block outputs.
+// A prim::BlockExit(*vals) denotes that the values are targeting the
+// block the Exit is contained on, and a prim::LoopExit(*vals) denotes that
+// the values are targeting the most recent loop block. FunctionExits are NYI.
+// Once we hit an exit node, we do not execute any further instructions
+// until the block exit reaches its destination. If a node may hit an exit
+// node that pauses execution, we use a boolean value to indicate if the exit
+// has been hit or not, and conditionalize further execution.
+// First we remove block exits of if nodes, then we replace Loop Block exits
+// with LoopExits. Then we remove LoopExits.
+//   block0(%i.2 : int, %i.12 : int):
```

Comment:
Please name the variables in the example. Without the names it's really hard to follow what's going on.

Reply:
agreed, Michael and I tried to follow this example and we had to give up. Providing descriptive names or at least the definitions for all used values would be very helpful. Also, showing the same transformation in pseudo python might be helpful as well.

Reply:
Sure, I will include a psuedo python version as well